### ISSUE TITLE
QScript for Filter with Reduce or multiple projections attempts ProjectIndex excessively

### ISSUE THREAD

**Issue by wemrysi** [quasar-analytics/quasar#1622]
Queries involving a `Filter` and a `Reduce` or a `Filter` and multiple projections results in QScript that attempts to `ProjectIndex` too far into the data from the backend. Included is an example of the former as it is more focused (to trigger the latter, a query such as `select city, loc from /explain/data where pop > 15000` suffices, but the output is about twice as long).

Interestingly, projecting a single field or `*` does not suffer from this problem.

If I'm reading the QScript correctly, the initial `LeftShift` produces a dataset where each datum has the shape `[<id>, <value>]` which the `Repair` expression then attempts to `ProjectIndex(ProjectIndex(ProjectIndex(RightSide, 0), 0), 0)`. 

**`select count(*) from /explain/data where pop > 1000`**

```
[info] Logical Plan:
[info] Let('tmp0)
[info] ├─ Read(/explain/data)
[info] ╰─ Let('tmp1)
[info]    ├─ Invoke(WHERE)
[info]    │  ├─ Free('tmp0)
[info]    │  ╰─ Invoke((>))
[info]    │     ├─ ObjectProject('tmp0["pop"])
[info]    │     ╰─ Constant(Int(1000))
[info]    ╰─ Let('tmp2)
[info]       ├─ Invoke(SQUASH)
[info]       │  ╰─ Invoke(MAKE_OBJECT)
[info]       │     ├─ Str("0")
[info]       │     ╰─ Invoke(COUNT)
[info]       │        ╰─ Free('tmp1)
[info]       ╰─ Free('tmp2)
[info]
[info] Optimized:
[info] Let('__tmp0)
[info] ├─ Read(/explain/data)
[info] ╰─ Invoke(SQUASH)
[info]    ╰─ Invoke(MAKE_OBJECT)
[info]       ├─ Str("0")
[info]       ╰─ Invoke(COUNT)
[info]          ╰─ Invoke(WHERE)
[info]             ├─ Free('__tmp0)
[info]             ╰─ Invoke((>))
[info]                ├─ ObjectProject('__tmp0["pop"])
[info]                ╰─ Constant(Int(1000))
[info]
[info] Typechecked:
[info] Let('__tmp0)
[info] ├─ Let('__check0)
[info] │  ├─ Read(/explain/data)
[info] │  ╰─ Typecheck(Obj(Map[], Some(Top)))
[info] │     ├─ Free('__check0)
[info] │     ├─ Free('__check0)
[info] │     ╰─ Constant(NA)
[info] ╰─ Invoke(SQUASH)
[info]    ╰─ Invoke(MAKE_OBJECT)
[info]       ├─ Str("0")
[info]       ╰─ Invoke(COUNT)
[info]          ╰─ Invoke(WHERE)
[info]             ├─ Free('__tmp0)
[info]             ╰─ Let('__check1)
[info]                ├─ ObjectProject('__tmp0["pop"])
[info]                ╰─ Typecheck(Int, Dec, Interval, Str, Timestamp, Date, Time, or Bool)
[info]                   ├─ Free('__check1)
[info]                   ├─ Invoke((>))
[info]                   │  ├─ Free('__check1)
[info]                   │  ╰─ Constant(Int(1000))
[info]                   ╰─ Constant(NA)
[info] QScript:
[info] Reduce
[info] ├─ Filter
[info] │  ├─ LeftShift
[info] │  │  ├─ Read(/quasar-test/run_a4acd653/explain/data)
[info] │  │  ├─ Struct
[info] │  │  │  ╰─ ZipMapKeys
[info] │  │  │     ╰─ ○
[info] │  │  ╰─ Repair
[info] │  │     ╰─ ConcatArrays
[info] │  │        ├─ ConcatArrays
[info] │  │        │  ├─ ConcatArrays
[info] │  │        │  │  ├─ MakeArray
[info] │  │        │  │  │  ╰─ ProjectIndex
[info] │  │        │  │  │     ├─ ProjectIndex
[info] │  │        │  │  │     │  ├─ ProjectIndex
[info] │  │        │  │  │     │  │  ├─ RightSide
[info] │  │        │  │  │     │  │  ╰─ Constant(Int(0))
[info] │  │        │  │  │     │  ╰─ Constant(Int(0))
[info] │  │        │  │  │     ╰─ Constant(Int(0))
[info] │  │        │  │  ╰─ MakeArray
[info] │  │        │  │     ╰─ ConcatArrays
[info] │  │        │  │        ├─ MakeArray
[info] │  │        │  │        │  ╰─ ProjectIndex
[info] │  │        │  │        │     ├─ ProjectIndex
[info] │  │        │  │        │     │  ├─ ProjectIndex
[info] │  │        │  │        │     │  │  ├─ ProjectIndex
[info] │  │        │  │        │     │  │  │  ├─ ProjectIndex
[info] │  │        │  │        │     │  │  │  │  ├─ RightSide
[info] │  │        │  │        │     │  │  │  │  ╰─ Constant(Int(0))
[info] │  │        │  │        │     │  │  │  ╰─ Constant(Int(0))
[info] │  │        │  │        │     │  │  ╰─ Constant(Int(0))
[info] │  │        │  │        │     │  ╰─ Constant(Int(0))
[info] │  │        │  │        │     ╰─ Constant(Int(0))
[info] │  │        │  │        ╰─ MakeArray
[info] │  │        │  │           ╰─ ProjectIndex
[info] │  │        │  │              ├─ ProjectIndex
[info] │  │        │  │              │  ├─ ProjectIndex
[info] │  │        │  │              │  │  ├─ ProjectIndex
[info] │  │        │  │              │  │  │  ├─ ProjectIndex
[info] │  │        │  │              │  │  │  │  ├─ ProjectIndex
[info] │  │        │  │              │  │  │  │  │  ├─ RightSide
[info] │  │        │  │              │  │  │  │  │  ╰─ Constant(Int(0))
[info] │  │        │  │              │  │  │  │  ╰─ Constant(Int(0))
[info] │  │        │  │              │  │  │  ╰─ Constant(Int(0))
[info] │  │        │  │              │  │  ╰─ Constant(Int(0))
[info] │  │        │  │              │  ╰─ Constant(Int(0))
[info] │  │        │  │              ╰─ Constant(Int(0))
[info] │  │        │  ╰─ MakeArray
[info] │  │        │     ╰─ Guard
[info] │  │        │        ├─ ProjectIndex
[info] │  │        │        │  ├─ RightSide
[info] │  │        │        │  ╰─ Constant(Int(1))
[info] │  │        │        ├─ Type(Obj(Map[], Some(Top)))
[info] │  │        │        ├─ ProjectIndex
[info] │  │        │        │  ├─ RightSide
[info] │  │        │        │  ╰─ Constant(Int(1))
[info] │  │        │        ╰─ Undefined
[info] │  │        ╰─ MakeArray
[info] │  │           ╰─ Guard
[info] │  │              ├─ ProjectField
[info] │  │              │  ├─ Guard
[info] │  │              │  │  ├─ ProjectIndex
[info] │  │              │  │  │  ├─ RightSide
[info] │  │              │  │  │  ╰─ Constant(Int(1))
[info] │  │              │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │  │              │  │  ├─ ProjectIndex
[info] │  │              │  │  │  ├─ RightSide
[info] │  │              │  │  │  ╰─ Constant(Int(1))
[info] │  │              │  │  ╰─ Undefined
[info] │  │              │  ╰─ Constant(Str(pop))
[info] │  │              ├─ Type(Int, Dec, Interval, Str, Timestamp, Date, Time, or Bool)
[info] │  │              ├─ Gt
[info] │  │              │  ├─ ProjectField
[info] │  │              │  │  ├─ Guard
[info] │  │              │  │  │  ├─ ProjectIndex
[info] │  │              │  │  │  │  ├─ RightSide
[info] │  │              │  │  │  │  ╰─ Constant(Int(1))
[info] │  │              │  │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │  │              │  │  │  ├─ ProjectIndex
[info] │  │              │  │  │  │  ├─ RightSide
[info] │  │              │  │  │  │  ╰─ Constant(Int(1))
[info] │  │              │  │  │  ╰─ Undefined
[info] │  │              │  │  ╰─ Constant(Str(pop))
[info] │  │              │  ╰─ Constant(Int(1000))
[info] │  │              ╰─ Undefined
[info] │  ╰─ ProjectIndex
[info] │     ├─ ○
[info] │     ╰─ Constant(Int(3))
[info] ├─ Bucket
[info] │  ╰─ Constant(Null())
[info] ├─ Reducers
[info] │  ╰─ Count
[info] │     ╰─ ProjectIndex
[info] │        ├─ ○
[info] │        ╰─ Constant(Int(2))
[info] ╰─ Repair
[info]    ╰─ MakeMap
[info]       ├─ Constant(Str(0))
[info]       ╰─ ReduceIndex(0)

[info] QScript (ShiftRead):
[info] List
[info] ├─ ShiftedRead(ShiftedRead(/quasar-test/run_a4acd653/explain/data, IncludeId))
[info] ├─ Map
[info] │  ├─ ExternallyManaged(Extern)
[info] │  ╰─ ConcatArrays
[info] │     ├─ ConcatArrays
[info] │     │  ├─ ConcatArrays
[info] │     │  │  ├─ MakeArray
[info] │     │  │  │  ╰─ ProjectIndex
[info] │     │  │  │     ├─ ProjectIndex
[info] │     │  │  │     │  ├─ ProjectIndex
[info] │     │  │  │     │  │  ├─ ConcatArrays
[info] │     │  │  │     │  │  │  ├─ MakeArray
[info] │     │  │  │     │  │  │  │  ╰─ ProjectIndex
[info] │     │  │  │     │  │  │  │     ├─ ○
[info] │     │  │  │     │  │  │  │     ╰─ Constant(Int(0))
[info] │     │  │  │     │  │  │  ╰─ MakeArray
[info] │     │  │  │     │  │  │     ╰─ ProjectIndex
[info] │     │  │  │     │  │  │        ├─ ○
[info] │     │  │  │     │  │  │        ╰─ Constant(Int(1))
[info] │     │  │  │     │  │  ╰─ Constant(Int(0))
[info] │     │  │  │     │  ╰─ Constant(Int(0))
[info] │     │  │  │     ╰─ Constant(Int(0))
[info] │     │  │  ╰─ MakeArray
[info] │     │  │     ╰─ ConcatArrays
[info] │     │  │        ├─ MakeArray
[info] │     │  │        │  ╰─ ProjectIndex
[info] │     │  │        │     ├─ ProjectIndex
[info] │     │  │        │     │  ├─ ProjectIndex
[info] │     │  │        │     │  │  ├─ ProjectIndex
[info] │     │  │        │     │  │  │  ├─ ProjectIndex
[info] │     │  │        │     │  │  │  │  ├─ ConcatArrays
[info] │     │  │        │     │  │  │  │  │  ├─ MakeArray
[info] │     │  │        │     │  │  │  │  │  │  ╰─ ProjectIndex
[info] │     │  │        │     │  │  │  │  │  │     ├─ ○
[info] │     │  │        │     │  │  │  │  │  │     ╰─ Constant(Int(0))
[info] │     │  │        │     │  │  │  │  │  ╰─ MakeArray
[info] │     │  │        │     │  │  │  │  │     ╰─ ProjectIndex
[info] │     │  │        │     │  │  │  │  │        ├─ ○
[info] │     │  │        │     │  │  │  │  │        ╰─ Constant(Int(1))
[info] │     │  │        │     │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  ╰─ Constant(Int(0))
[info] │     │  │        │     ╰─ Constant(Int(0))
[info] │     │  │        ╰─ MakeArray
[info] │     │  │           ╰─ ProjectIndex
[info] │     │  │              ├─ ProjectIndex
[info] │     │  │              │  ├─ ProjectIndex
[info] │     │  │              │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  │  │  ├─ ConcatArrays
[info] │     │  │              │  │  │  │  │  │  ├─ MakeArray
[info] │     │  │              │  │  │  │  │  │  │  ╰─ ProjectIndex
[info] │     │  │              │  │  │  │  │  │  │     ├─ ○
[info] │     │  │              │  │  │  │  │  │  │     ╰─ Constant(Int(0))
[info] │     │  │              │  │  │  │  │  │  ╰─ MakeArray
[info] │     │  │              │  │  │  │  │  │     ╰─ ProjectIndex
[info] │     │  │              │  │  │  │  │  │        ├─ ○
[info] │     │  │              │  │  │  │  │  │        ╰─ Constant(Int(1))
[info] │     │  │              │  │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  ╰─ Constant(Int(0))
[info] │     │  │              ╰─ Constant(Int(0))
[info] │     │  ╰─ MakeArray
[info] │     │     ╰─ Guard
[info] │     │        ├─ ProjectIndex
[info] │     │        │  ├─ ConcatArrays
[info] │     │        │  │  ├─ MakeArray
[info] │     │        │  │  │  ╰─ ProjectIndex
[info] │     │        │  │  │     ├─ ○
[info] │     │        │  │  │     ╰─ Constant(Int(0))
[info] │     │        │  │  ╰─ MakeArray
[info] │     │        │  │     ╰─ ProjectIndex
[info] │     │        │  │        ├─ ○
[info] │     │        │  │        ╰─ Constant(Int(1))
[info] │     │        │  ╰─ Constant(Int(1))
[info] │     │        ├─ Type(Obj(Map[], Some(Top)))
[info] │     │        ├─ ProjectIndex
[info] │     │        │  ├─ ConcatArrays
[info] │     │        │  │  ├─ MakeArray
[info] │     │        │  │  │  ╰─ ProjectIndex
[info] │     │        │  │  │     ├─ ○
[info] │     │        │  │  │     ╰─ Constant(Int(0))
[info] │     │        │  │  ╰─ MakeArray
[info] │     │        │  │     ╰─ ProjectIndex
[info] │     │        │  │        ├─ ○
[info] │     │        │  │        ╰─ Constant(Int(1))
[info] │     │        │  ╰─ Constant(Int(1))
[info] │     │        ╰─ Undefined
[info] │     ╰─ MakeArray
[info] │        ╰─ Guard
[info] │           ├─ ProjectField
[info] │           │  ├─ Guard
[info] │           │  │  ├─ ProjectIndex
[info] │           │  │  │  ├─ ConcatArrays
[info] │           │  │  │  │  ├─ MakeArray
[info] │           │  │  │  │  │  ╰─ ProjectIndex
[info] │           │  │  │  │  │     ├─ ○
[info] │           │  │  │  │  │     ╰─ Constant(Int(0))
[info] │           │  │  │  │  ╰─ MakeArray
[info] │           │  │  │  │     ╰─ ProjectIndex
[info] │           │  │  │  │        ├─ ○
[info] │           │  │  │  │        ╰─ Constant(Int(1))
[info] │           │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │           │  │  ├─ ProjectIndex
[info] │           │  │  │  ├─ ConcatArrays
[info] │           │  │  │  │  ├─ MakeArray
[info] │           │  │  │  │  │  ╰─ ProjectIndex
[info] │           │  │  │  │  │     ├─ ○
[info] │           │  │  │  │  │     ╰─ Constant(Int(0))
[info] │           │  │  │  │  ╰─ MakeArray
[info] │           │  │  │  │     ╰─ ProjectIndex
[info] │           │  │  │  │        ├─ ○
[info] │           │  │  │  │        ╰─ Constant(Int(1))
[info] │           │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  ╰─ Undefined
[info] │           │  ╰─ Constant(Str(pop))
[info] │           ├─ Type(Int, Dec, Interval, Str, Timestamp, Date, Time, or Bool)
[info] │           ├─ Gt
[info] │           │  ├─ ProjectField
[info] │           │  │  ├─ Guard
[info] │           │  │  │  ├─ ProjectIndex
[info] │           │  │  │  │  ├─ ConcatArrays
[info] │           │  │  │  │  │  ├─ MakeArray
[info] │           │  │  │  │  │  │  ╰─ ProjectIndex
[info] │           │  │  │  │  │  │     ├─ ○
[info] │           │  │  │  │  │  │     ╰─ Constant(Int(0))
[info] │           │  │  │  │  │  ╰─ MakeArray
[info] │           │  │  │  │  │     ╰─ ProjectIndex
[info] │           │  │  │  │  │        ├─ ○
[info] │           │  │  │  │  │        ╰─ Constant(Int(1))
[info] │           │  │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │           │  │  │  ├─ ProjectIndex
[info] │           │  │  │  │  ├─ ConcatArrays
[info] │           │  │  │  │  │  ├─ MakeArray
[info] │           │  │  │  │  │  │  ╰─ ProjectIndex
[info] │           │  │  │  │  │  │     ├─ ○
[info] │           │  │  │  │  │  │     ╰─ Constant(Int(0))
[info] │           │  │  │  │  │  ╰─ MakeArray
[info] │           │  │  │  │  │     ╰─ ProjectIndex
[info] │           │  │  │  │  │        ├─ ○
[info] │           │  │  │  │  │        ╰─ Constant(Int(1))
[info] │           │  │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  │  ╰─ Undefined
[info] │           │  │  ╰─ Constant(Str(pop))
[info] │           │  ╰─ Constant(Int(1000))
[info] │           ╰─ Undefined
[info] ├─ Filter
[info] │  ├─ ExternallyManaged(Extern)
[info] │  ╰─ ProjectIndex
[info] │     ├─ ○
[info] │     ╰─ Constant(Int(3))
[info] ╰─ Reduce
[info]    ├─ ExternallyManaged(Extern)
[info]    ├─ Bucket
[info]    │  ╰─ Constant(Null())
[info]    ├─ Reducers
[info]    │  ╰─ Count
[info]    │     ╰─ ProjectIndex
[info]    │        ├─ ○
[info]    │        ╰─ Constant(Int(2))
[info]    ╰─ Repair
[info]       ╰─ MakeMap
[info]          ├─ Constant(Str(0))
[info]          ╰─ ReduceIndex(0)

[info] QScript (Optimized):
[info] List
[info] ├─ ShiftedRead(ShiftedRead(/quasar-test/run_a4acd653/explain/data, IncludeId))
[info] ├─ Map
[info] │  ├─ ExternallyManaged(Extern)
[info] │  ╰─ ConcatArrays
[info] │     ├─ ConcatArrays
[info] │     │  ├─ ConcatArrays
[info] │     │  │  ├─ MakeArray
[info] │     │  │  │  ╰─ ProjectIndex
[info] │     │  │  │     ├─ ProjectIndex
[info] │     │  │  │     │  ├─ ProjectIndex
[info] │     │  │  │     │  │  ├─ ○
[info] │     │  │  │     │  │  ╰─ Constant(Int(0))
[info] │     │  │  │     │  ╰─ Constant(Int(0))
[info] │     │  │  │     ╰─ Constant(Int(0))
[info] │     │  │  ╰─ MakeArray
[info] │     │  │     ╰─ ConcatArrays
[info] │     │  │        ├─ MakeArray
[info] │     │  │        │  ╰─ ProjectIndex
[info] │     │  │        │     ├─ ProjectIndex
[info] │     │  │        │     │  ├─ ProjectIndex
[info] │     │  │        │     │  │  ├─ ProjectIndex
[info] │     │  │        │     │  │  │  ├─ ProjectIndex
[info] │     │  │        │     │  │  │  │  ├─ ○
[info] │     │  │        │     │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  │  ╰─ Constant(Int(0))
[info] │     │  │        │     │  ╰─ Constant(Int(0))
[info] │     │  │        │     ╰─ Constant(Int(0))
[info] │     │  │        ╰─ MakeArray
[info] │     │  │           ╰─ ProjectIndex
[info] │     │  │              ├─ ProjectIndex
[info] │     │  │              │  ├─ ProjectIndex
[info] │     │  │              │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  │  ├─ ProjectIndex
[info] │     │  │              │  │  │  │  │  ├─ ○
[info] │     │  │              │  │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  │  ╰─ Constant(Int(0))
[info] │     │  │              │  ╰─ Constant(Int(0))
[info] │     │  │              ╰─ Constant(Int(0))
[info] │     │  ╰─ MakeArray
[info] │     │     ╰─ Guard
[info] │     │        ├─ ProjectIndex
[info] │     │        │  ├─ ○
[info] │     │        │  ╰─ Constant(Int(1))
[info] │     │        ├─ Type(Obj(Map[], Some(Top)))
[info] │     │        ├─ ProjectIndex
[info] │     │        │  ├─ ○
[info] │     │        │  ╰─ Constant(Int(1))
[info] │     │        ╰─ Undefined
[info] │     ╰─ MakeArray
[info] │        ╰─ Guard
[info] │           ├─ ProjectField
[info] │           │  ├─ Guard
[info] │           │  │  ├─ ProjectIndex
[info] │           │  │  │  ├─ ○
[info] │           │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │           │  │  ├─ ProjectIndex
[info] │           │  │  │  ├─ ○
[info] │           │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  ╰─ Undefined
[info] │           │  ╰─ Constant(Str(pop))
[info] │           ├─ Type(Int, Dec, Interval, Str, Timestamp, Date, Time, or Bool)
[info] │           ├─ Gt
[info] │           │  ├─ ProjectField
[info] │           │  │  ├─ Guard
[info] │           │  │  │  ├─ ProjectIndex
[info] │           │  │  │  │  ├─ ○
[info] │           │  │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  │  ├─ Type(Obj(Map[], Some(Top)))
[info] │           │  │  │  ├─ ProjectIndex
[info] │           │  │  │  │  ├─ ○
[info] │           │  │  │  │  ╰─ Constant(Int(1))
[info] │           │  │  │  ╰─ Undefined
[info] │           │  │  ╰─ Constant(Str(pop))
[info] │           │  ╰─ Constant(Int(1000))
[info] │           ╰─ Undefined
[info] ├─ Filter
[info] │  ├─ ExternallyManaged(Extern)
[info] │  ╰─ ProjectIndex
[info] │     ├─ ○
[info] │     ╰─ Constant(Int(3))
[info] ╰─ Reduce
[info]    ├─ ExternallyManaged(Extern)
[info]    ├─ Bucket
[info]    │  ╰─ Constant(Null())
[info]    ├─ Reducers
[info]    │  ╰─ Count
[info]    │     ╰─ ProjectIndex
[info]    │        ├─ ○
[info]    │        ╰─ Constant(Int(2))
[info]    ╰─ Repair
[info]       ╰─ MakeMap
[info]          ├─ Constant(Str(0))
[info]          ╰─ ReduceIndex(0)
```

**Comment by jdegoes** [comment]
@sellout @alissapajer Looks bad, I put this one in ASAP.

**Comment by wemrysi** [comment]
@sellout @alissapajer As an update, fixing the MarkLogic planner to preserve `Undefined` values in arrays does allow the query to execute correctly.

However, maintaining all the extra, unused information has a pretty big performance impact currently. The XQuery generated by the above plan takes ~28.5s to execute whereas if I take the same XQuery and just remove the `Map` that creates all the additional information (and update the `Filter` predicate appropriately) the result executes in ~2.6s. I'm sure some of this is specific to MarkLogic (i.e. not all connectors will slow down this much), but it will definitely be nice once this is elided. I'm going to see if leveraging the `StaticArray` and `StaticMap` extractors speeds things up by avoiding all the make/concat operations.

**Comment by alissapajer** [comment]
@wemrysi good to hear that the query is executing correctly, but obviously the performance is not acceptable. Thanks for the update.

**Comment by alissapajer** [comment]
The main example provided above will be indirectly addressed by #1745 because the excessively projected indices will be pruned since they are not used. But this bug will still be hit in queries where they are used. For example:

```scala
  "convert distinct by" in {
    val lp = fullCompileExp("select distinct(city) from zips order by pop")
    val qs = convert(listContents.some, lp)
    qs must beSome(beQScript(chain(
      ReadR(rootDir </> file("zips")),
      QC.inj(LeftShift((),
        HoleF,
        IncludeId,
        Free.roll(ConcatArrays(
          Free.roll(ConcatArrays(
            // FIXME: #1622
            Free.roll(ConcatArrays(
              Free.roll(MakeArray(
                ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(RightSideF, IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)))),
              Free.roll(MakeArray(
                ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(ProjectIndexR(RightSideF, IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)), IntLit(0)))))),
            Free.roll(MakeArray(
              Free.roll(ConcatMaps(
                Free.roll(MakeMap(
                  StrLit("city"),
                  ProjectFieldR(
                    Free.roll(Guard(
                      ProjectIndexR(RightSideF, IntLit(1)),
                      Type.Obj(ScalaMap(),Some(Type.Top)),
                      ProjectIndexR(RightSideF, IntLit(1)),
                      Free.roll(Undefined()))),
                    StrLit("city")))),
                Free.roll(MakeMap(
                  StrLit("__sd__0"),
                  ProjectFieldR(
                    Free.roll(Guard(
                      ProjectIndexR(RightSideF, IntLit(1)),
                      Type.Obj(ScalaMap(),Some(Type.Top)),
                      ProjectIndexR(RightSideF, IntLit(1)),
                      Free.roll(Undefined()))),
                    StrLit("pop")))))))))),
          Free.roll(MakeArray(
            ProjectFieldR(
              Free.roll(Guard(
                ProjectIndexR(RightSideF, IntLit(1)),
                Type.Obj(ScalaMap(),Some(Type.Top)),
                ProjectIndexR(RightSideF, IntLit(1)),
                Free.roll(Undefined()))),
              StrLit("pop")))))))),
      QC.inj(Sort((),
        Free.roll(ConcatArrays(
          Free.roll(MakeArray(ProjectIndexR(ProjectIndexR(HoleF, IntLit(0)), IntLit(0)))),
          Free.roll(MakeArray(ProjectIndexR(ProjectIndexR(HoleF, IntLit(1)), IntLit(0)))))),
        (ProjectIndexR(HoleF, IntLit(3)) -> SortDir.asc).wrapNel)),
      QC.inj(Reduce((),
        Free.roll(DeleteField(ProjectIndexR(HoleF, IntLit(2)), StrLit("__sd__0"))),
        List(ReduceFuncs.Arbitrary(ProjectIndexR(HoleF, IntLit(2)))),
        Free.roll(DeleteField(ReduceIndexF(0), StrLit("__sd__0"))))))))
}
```

**Comment by alissapajer** [comment]
Closed with merge of #1767
